### PR TITLE
Refactor MsgFormat into ROS1 and ROS2 versions.

### PR DESCRIPTION
**Pull request by schmerl**
Related to #420

### CODE REVIEW

File: src/roswire/common/msg.py [hunk 1]
```diff
@@ -245,10 +253,10 @@ def toposort(fmts: Collection["MsgFormat"]) -> List["MsgFormat"]:
             raise exc.PackageNotFound(missing_package_name)
         return [fn_to_fmt[filename] for filename in toposorted]
 
-    @staticmethod
-    def from_file(
-        package: str, filename: str, files: dockerblade.FileSystem
-    ) -> "MsgFormat":
+    @classmethod
+    def from_file(cls,
+                  package: str, filename: str, files: dockerblade.FileSystem
```
Comment (ChrisTimperley) [inline src/roswire/common/msg.py:258]: whitespace

File: src/roswire/common/msg.py [hunk 2]
```diff
@@ -345,10 +335,10 @@ def fullname(self) -> str:
         return f"{self.package}/{self.name}"
 
     def flatten(
-        self,
-        name_to_format: Mapping[str, "MsgFormat"],
-        ctx: Tuple[str, ...] = (),
-    ) -> Iterator[Tuple[Tuple[str, ...], Field]]:
+            self,
```
Comment (ChrisTimperley) [inline src/roswire/common/msg.py:338]: whitespace

File: src/roswire/common/msg.py [hunk 3]
```diff
@@ -378,6 +368,9 @@ def md5sum(self, name_to_msg: Mapping[str, "MsgFormat"]) -> str:
         return md5sum
 
 
+MF = TypeVar("MF", bound='MsgFormat')
```
Comment (ChrisTimperley) [inline src/roswire/common/msg.py:371]: Where is this used?

File: src/roswire/ros2/format.py [hunk 4]
```diff
@@ -4,21 +4,21 @@
 from typing import Any, Dict
 
 from .action import ROS2ActionFormat
+from .msg import ROS2MsgFormat
 from .srv import ROS2SrvFormat
 from ..common import (
```
Comment (ChrisTimperley) [inline src/roswire/ros2/format.py:9]: no needs for parens

File: src/roswire/ros2/msg.py [hunk 5]
```diff
@@ -2,73 +2,74 @@
...
 class ROS2Field(Field):
     R_DEFAULT_VALUE = r"[^#]*"
-    R_FIELD = re.compile(f"^\s*({Field.R_TYPE})(\s+)({Field.R_NAME})"
-                         f"\s+({R_DEFAULT_VALUE})\s*"
-                         f"{R_COMMENT}$")
+    R_FIELD = re.compile(f"^\s*(?P<type>{Field.R_TYPE})"
+                         f"\s+(?P<name>{Field.R_NAME})(?:\s+)?"
+                         f"(?P<val>{R_DEFAULT_VALUE}){R_COMMENT}")
 
-    default_val: Optional[str]
```
Comment (ChrisTimperley) [inline src/roswire/ros2/msg.py:21]: Is this a safe change? Are there clients that expect Field to have a default_val attribute?

File: src/roswire/ros2/msg.py [hunk 6]
```diff
@@ -2,73 +2,74 @@
...
-
-            field: ROS2Field = ROS2Field(old_field.typ,
-                                         old_field.name,
-                                         def_val)
+        if m_field:
+            typ = m_field.group('type')
+            name = m_field.group('name')
+            typ = cls._resolve_type(package, typ)
+            val = m_field.group('val')
+            field = ROS2Field(typ, name, val if val and val != '' else None)
```
Comment (ChrisTimperley) [inline src/roswire/ros2/msg.py:32]: `val` cannot be an empty string if it evaluates to `False`, so just `if val else` is equivalent.

File: src/roswire/ros2/msg.py [hunk 7]
```diff
@@ -2,73 +2,74 @@
...
         return ROS2Field.from_string(package, line)
 
+    @classmethod
+    def _field_from_dict(cls, dict: Dict[str, Any]) -> Field:
+        return ROS2Field.from_dict(dict)
+
     @classmethod
     def from_file(
-        cls, package: str, filename: str, files: dockerblade.FileSystem
+            cls, package: str, filename: str, files: dockerblade.FileSystem
```
Comment (ChrisTimperley) [inline src/roswire/ros2/msg.py:72]: whitespace

File: src/roswire/ros2/package.py [hunk 8]
```diff
@@ -18,10 +18,10 @@
 from loguru import logger
 from typing_extensions import Final
 
+from .action import ROS2ActionFormat
+from .msg import ROS2MsgFormat
 from .srv import ROS2SrvFormat
-from ..common import (ActionFormat,
-                      MsgFormat,
-                      Package,
+from ..common import (Package,
```
Comment (ChrisTimperley) [inline src/roswire/ros2/package.py:24]: collapse into one line

File: src/roswire/common/msg.py [hunk 9]
```diff
@@ -4,19 +4,20 @@
...
+from abc import ABC, abstractmethod
 from io import BytesIO
 from typing import (
     Any,
     BinaryIO,
-    ClassVar,
     Collection,
     Dict,
-    Iterator,
+    Generic, Iterator,
```
Comment (ChrisTimperley) [inline src/roswire/common/msg.py:14]: split across two lines

File: src/roswire/common/msg.py [hunk 10]
```diff
@@ -290,42 +299,23 @@ def from_string(package: str, name: str, text: str) -> "MsgFormat":
...
-    @staticmethod
-    def from_dict(
-        d: Dict[str, Any],
-        *,
-        package: Optional[str] = None,
-        name: Optional[str] = None,
-    ) -> "MsgFormat":
+    @classmethod
+    def from_dict(cls,
+                  d: Dict[str, Any],
```
Comment (ChrisTimperley) [inline src/roswire/common/msg.py:306]: keep whitespace/indentation consistent

File: src/roswire/ros2/msg.py (hunk 5 above)
Comment (schmerl) [reply src/roswire/ros2/msg.py:21]: Not sure what you mean. It though this was the culmination of what the changes has been about - to allow fields to have a default_val.

File: src/roswire/ros2/msg.py [hunk 11]
```diff
@@ -2,73 +2,74 @@
...
+                constants.append(constant)
+            elif field:
+                fields.append(field)
+            else:
+                raise ParsingError(f"failed to parse line: {line}")
 
-    @classmethod
-    def _field_from_dict(cls, dict: Dict[str, Any]) -> ROS2Field:
-        return ROS2Field.from_dict(dict)
+        return ROS2MsgFormat(package, name, text, fields, constants)  # type: ignore  # noqa
```
[the remark this answers is not in the document]
Comment (schmerl) [reply src/roswire/ros2/msg.py:60]: Inherited from master branch, MsgFormat.

File: src/roswire/ros2/msg.py (hunk 5 above)
Comment (ChrisTimperley) [reply src/roswire/ros2/msg.py:21]: Suggest rename to `default_value`.

Comment (ChrisTimperley) [reply src/roswire/ros2/msg.py:21]: It looks like this hasn't been changed yet?

Comment (schmerl) [reply src/roswire/ros2/msg.py:21]: I'm very confused about this. If I go to the latest version of the file I see that it is `default_value`. It just doesn't appear in this comment,

**Review by ChrisTimperley — APPROVED** [review]

**Pull request merged**